src/game/carrier.c

```diff
@@ -234,8 +273,68 @@ void Carrier_TestLegacyDrops(int16_t item_num)
         CARRIED_ITEM *item = carrier->carried_item;
         while (item) {
             // Simulate Lara having picked up the item.
-            item->status = IS_INVISIBLE;
+            item->status = DS_COLLECTED;
             item = item->next_item;
         }
     }
 }
+
+void Carrier_AnimateDrops(void)
+{
+    if (!m_AnimatingCount) {
+        return;
+    }
+
+    // Make items that spawn in mid-air or water gracefully fall to the floor.
+    for (int i = 0; i < g_LevelItemCount; i++) {
+        ITEM_INFO *carrier = &g_Items[i];
+        CARRIED_ITEM *item = carrier->carried_item;
+        if (!item) {
+            continue;
+        }
+
+        do {
+            if (item->status != DS_FALLING) {
+                continue;
+            }
+
+            ITEM_INFO *pickup = &g_Items[item->spawn_number];
+            int16_t room_num = pickup->room_number;
+            FLOOR_INFO *floor = Room_GetFloor(
+                pickup->pos.x, pickup->pos.y, pickup->pos.z, &room_num);
+            int16_t height = Room_GetHeight(
+                floor, pickup->pos.x, pickup->pos.y, pickup->pos.z);
+            bool in_water =
+                g_RoomInfo[pickup->room_number].flags & RF_UNDERWATER;
+
+            if (floor->pit_room == NO_ROOM && pickup->pos.y >= height) {
+                item->status = DS_DROPPED;
+                pickup->pos.y = height;
+                pickup->fall_speed = 0;
+                m_AnimatingCount--;
+            } else {
+                pickup->fall_speed +=
+                    (!in_water && pickup->fall_speed < FASTFALL_SPEED)
+                    ? DROP_FAST_RATE
+                    : DROP_SLOW_RATE;
+                pickup->pos.y += pickup->fall_speed;
+                pickup->pos.y_rot += in_water ? DROP_SLOW_TURN : DROP_FAST_TURN;
+
+                if (floor->pit_room != NO_ROOM
+                    && pickup->pos.y > (floor->floor << 8)) {
+                    room_num = floor->pit_room;
+                }
+            }
+
+            if (room_num != pickup->room_number) {
+                Item_NewRoom(item->spawn_number, room_num);
+            }
+
+            // Track animating status in the carrier for saving/loading.
+            item->pos = pickup->pos;
+            item->room_number = pickup->room_number;
+            item->fall_speed = pickup->fall_speed;
+
+        } while ((item = item->next_item));
```

Comment:
Can we move the body of this loop to a separate static function to reduce indentation by 8 spaces?